### ISSUE TITLE
Build service images inside dev-box failed with 'Permission denied'

### ISSUE THREAD

**Issue by hao1939** [Microsoft/pai#710]
Reproduce step:

1. Start a dev-box with the command
`# Running your maintain-box as a docker container.
sudo docker run -itd \
        -e COLUMNS=$COLUMNS -e LINES=$LINES -e TERM=$TERM \
        -v /pathConfiguration:/cluster-configuration  \
        -v /var/lib/docker:/var/lib/docker \
        -v /var/run/docker.sock:/var/run/docker.sock \
        -v /pathHadoop:/pathHadoop \
        --pid=host \
        --privileged=true \
        --net=host \
        --name=dev-box \
        dev-box`

2. Access dev-box
`sudo docker exec -it dev-box /bin/bash`

3. cd /pai/pai-management
4. ./paictl.py image build -p /cluster-configuration

Then would encountered the error while build the 'end-to-end-test' image.

> Get:15 http://security.ubuntu.com/ubuntu xenial-security/multiverse amd64 Packages [3735 B]
Fetched 3656 kB in 4s (858 kB/s)
Reading package lists...
W: Problem unlinking the file /var/lib/apt/lists/partial/.apt-acquire-privs-test.424P1s - IsAccessibleBySandboxUser (13: Permission denied)
W: Problem unlinking the file /var/lib/apt/lists/partial/.apt-acquire-privs-test.NhdXW1 - IsAccessibleBySandboxUser (13: Permission denied)
W: Problem unlinking the file /var/lib/apt/lists/partial/.apt-acquire-privs-test.mFY4RA - IsAccessibleBySandboxUser (13: Permission denied)
W: Problem unlinking the file /var/lib/apt/lists/partial/.apt-acquire-privs-test.LHedN9 - IsAccessibleBySandboxUser (13: Permission denied)
E: Failed to stat /var/lib/apt/lists/partial/archive.ubuntu.com_ubuntu_dists_xenial-updates_universe_source_Sources.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/archive.ubuntu.com_ubuntu_dists_xenial-updates_main_binary-amd64_Packages.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/archive.ubuntu.com_ubuntu_dists_xenial-updates_restricted_binary-amd64_Packages.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/archive.ubuntu.com_ubuntu_dists_xenial-updates_universe_binary-amd64_Packages.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/archive.ubuntu.com_ubuntu_dists_xenial-updates_multiverse_binary-amd64_Packages.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/archive.ubuntu.com_ubuntu_dists_xenial-backports_universe_binary-amd64_Packages.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/security.ubuntu.com_ubuntu_dists_xenial-security_universe_source_Sources.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/security.ubuntu.com_ubuntu_dists_xenial-security_main_binary-amd64_Packages.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/security.ubuntu.com_ubuntu_dists_xenial-security_restricted_binary-amd64_Packages.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/security.ubuntu.com_ubuntu_dists_xenial-security_universe_binary-amd64_Packages.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
E: Failed to stat /var/lib/apt/lists/partial/security.ubuntu.com_ubuntu_dists_xenial-security_multiverse_binary-amd64_Packages.lz4 - pkgAcqTransactionItem::TransactionState-stat (2: No such file or directory)
The command '/bin/sh -c apt-get -y update &&     apt-get -y install python git jq &&     apt-get clean &&     rm -rf /var/lib/apt/lists/*' returned a non-zero code: 100
2018-06-12 05:58:01,209 [ERROR] - paiLibrary.common.linux_shell : An error occurs, when building your image [ end-to-end-test ]

**Comment by hao1939** [comment]
@ydye @abuccts ,
Could you take a look?

**Comment by hao1939** [comment]
Actually the error happens when execute: 

> 2018-06-12 05:57:55,116 [INFO] - paiLibrary.common.docker_handler : Begin to execute the command: docker build -t end-to-end-test ./src/end-to-end-test/

And docker build failed on the cmd below:
`/bin/sh -c apt-get -y update &&     apt-get -y install python git jq &&     apt-get clean &&     rm -rf /var/lib/apt/lists/*`

Could workaround by change the cmd to:
`apt-get -y install python git jq &&     rm -rf /var/lib/apt/lists/*`

We need to fix the base image.

**Comment by hao1939** [comment]
The same as #711 ,

Update the dockerfile and add one line before 'apt-get update' solved this issue:

```
RUN rm -rf /var/lib/apt/lists
RUN apt-get -y update && \
    apt-get -y install maven
```

**Comment by abuccts** [comment]
I don't think it's a good idea to clean up before `apt-get` in Dockerfile.
Do you build those images in a Docker instance? Please fix in the base Dockerfile instead.

**Comment by hao1939** [comment]
Good point!
I will take a try.

**Comment by hao1939** [comment]
Hi @abuccts ,
I noticed that we do clean up after `apt-get` in Dockerfile.
Is there any different between 'clean up in base Dockerfile after apt-get' or 'clean up in current Dockerfile before apt-get'?

Ps: I do agree we should fix the base, and we should make sure every image in the best situation.
Maybe we need to place the clean up in both.

**Comment by abuccts** [comment]
@hao1939 It's recommended to [clean up apt cache to reduce image size](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run), please check [base image](https://github.com/Microsoft/pai/blob/master/pai-management/src/base-image/dockerfile#L49) and [hadoop-run](https://github.com/Microsoft/pai/blob/master/pai-management/src/hadoop-run/dockerfile.template#L22).

What's the reason for cleaning cache before `apt-get`? Nothing has been installed so what should be cleaned up?

@ydye If it's necessary to build images inside a Docker instance, it's better to add `--no-cache` arg in `docker build`.

**Comment by hao1939** [comment]
@abuccts ,

Got you, it make sense. 

Here's my point.
Usually we put the cleaning cache before `apt-get` as a protection (for situation like this issue).
And place some redundancy make the Dockerfile more robust.

**Comment by abuccts** [comment]
Fixed in #714.
